openfl/federated/task/runner_tf.py

```diff
@@ -177,180 +133,350 @@ def train_batches(self, col_name, round_num, input_tensor_dict,
 
-    def reset_opt_vars(self):
-        """Reinitialize the optimizer variables."""
-        for v in self.opt_vars:
-            v.initializer.run(session=self.sess)
+        weight_values = [weights_dict[name] for name in weight_names]
 
-    def initialize_globals(self):
-        """Initialize Global Variables.
+        obj.set_weights(weight_values)
 
-        Initialize all global variables
+    def get_tensor_dict(self, with_opt_vars, suffix=''):
+        """
+        Get the model weights as a tensor dictionary.
+
+        Parameters
+        ----------
+        with_opt_vars : bool
+            If we should include the optimizer's status.
+        suffix : string
+            Universally
 
         Returns:
-            None
+            dict: The tensor dictionary.
         """
-        self.sess.run(tf.global_variables_initializer())
 
-    def _get_weights_names(self, with_opt_vars=True):
-        """Get the weights.
+        model_weights = self._get_weights_dict(self.model, suffix)
 
-        Args:
-            with_opt_vars (bool): Specify if we also want to get the variables
-             of the optimizer.
+        if with_opt_vars:
 
-        Returns:
-            list : The weight names list
+            opt_weights = self._get_weights_dict(self.model.optimizer, suffix, with_opt_vars)
+
+            model_weights.update(opt_weights)
+
+            if len(opt_weights) == 0:
+                self.logger.debug(
+                    "WARNING: We didn't find variables for the optimizer.")
+        return model_weights
+
+    def set_tensor_dict(self, tensor_dict, with_opt_vars):
         """
-        if with_opt_vars is True:
-            variables = self.fl_vars
+        Set the model weights with a tensor dictionary.
+
+        Args:
+            tensor_dict: the tensor dictionary
+            with_opt_vars (bool): True = include the optimizer's status.
+        """
+        if with_opt_vars is False:
+            # It is possible to pass in opt variables from the input tensor dict
+            # This will make sure that the correct layers are updated
+            model_weight_names = [weight.name for weight in self.model.weights]
+            model_weights_dict = {
+                name: tensor_dict[name] for name in model_weight_names
+            }
+            self._set_weights_dict(self.model, model_weights_dict)
         else:
-            variables = self.tvars
+            model_weight_names = [
+                weight.name for weight in self.model.weights
+            ]
+            model_weights_dict = {
+                name: tensor_dict[name] for name in model_weight_names
+            }
+
+            opt_weight_names = [
+                weight.name for weight in self.model.optimizer.variables
+            ]
+            opt_weights_dict = {
+                name: tensor_dict[name] for name in opt_weight_names
+            }
+            self._set_weights_dict(self.model, model_weights_dict)
```

Comment:
this statement can also be moved outside of the `if` statement as it's executed in both cases:
```python
self._set_weights_dict(self.model, model_weights_dict)
```